### PR TITLE
Make Faraday-APRS Persistent Across Network Failures

### PR DESCRIPTION
**Pull request by kb1lqc**
# BASE91 APRS Comment Telemetry
[aprs.fi](www.aprs.fi) and a few other services support BASE91 telemetry transmission. This allows up to 13 bits of ADC telemetry per channel. This PR pulled in what was #269 so here is the  information relevant to that PR. @reillyeon beat @el-iso to the punch on there. I appreciate the help from both!

## Changelog
- Imports aprslib and uses base91 functionality. Please note that this is note a standard BASE91 implementation and is a custom APRS version of base91. Qeue [XKCD](https://xkcd.com/927/).
- Removes use of `sendTelemetry` function as it is no longer needed
- Only sends scaling values on first transmission to reduce unnecessary packet transmissions. This should be revisited later but seems good for now.
- Sends a `telemetrySequence` number that is unbounded to `sendPosition` as BASE91 comment telemetry can support up to 16 bits of counting and this should be fine for the forceable future :)
- `sendPosition` implements [BASE91 Comment Telemetry](http://he.fi/doc/aprs-base91-comment-telemetry.txt) with ADC0, ADC1, ADC3, ADC6, BOARDTEMP, and GPIO bits per #266.

Raw packets which end in BASE91 comment telemetry 
image

12 bit ADC data looks very nice!
image

# Persistent APRS-IS Connection Across Network

This PR is based on #269 as it was branched from that PR repository and aims to improve reliability when connecting to APRS-IS servers across a network. As we've seen, often a few minutes to a few hours after starting APRS logging a network event will cause  socket ERRNO 10054 or similar and `faraday-aprs` could not recover. This was annoying at best and at worst rendered the program not terribly useful.

Mind helping me out @reillyeon, if no time I can see if @kb1lqd has some free time in the near future :)

## Changelog
- Created a `sendAPRSPacket` function to consolidate sending of data to the socket and exception catching
- Reconnects after a IOERROR in 2 second loop attempts

## Issue Tickets
This PR addresses issues related to the following tickets Faraday-Software #119, Faraday-Software #190, Faraday-Software #199, Faraday-Software #201, Faraday-Software #262. A few of these tickets can likely be completely closed as a result of this testing and eventual PR!

### CODE REVIEW

File: faraday/aprs.py [hunk 1]
```diff
@@ -103,6 +104,11 @@ def aprs_worker(config, sock):
     # Local variable initialization
     telemSequence = 0
 
+    #dev
+    sock = ''
```
Comment (reillyeon) [inline faraday/aprs.py]: This line is unnecessary.

File: faraday/aprs.py [hunk 2]
```diff
@@ -114,11 +120,21 @@ def aprs_worker(config, sock):
...
-        sendParameters(stationData, sock)
-        sendEquations(stationData, sock)
+        if not conn:
+            sock = connectAPRSIS()
+        try:
+            conn = sendPositions(telemSequence, stationData, sock)
+
+            # Just send labels, Parameters, and Equations every 10th loop
+            if telemSequence % 10 == 0:
+
```
Comment (reillyeon) [inline faraday/aprs.py]: nit: Extra blank line.

File: faraday/aprs.py (hunk 1 above)
Comment (reillyeon) [inline faraday/aprs.py]: nit: This statement is unnecessary.

File: faraday/aprs.py [hunk 3]
```diff
@@ -317,11 +382,8 @@ def sendPositions(stations, socket):
 
                 logger.debug(positionString)
 
-                try:
-                    socket.sendall(positionString)
-
-                except IOError as e:
-                    logger.error(e)
+                status = sendAPRSPacket(socket, positionString)
+                return status
```
Comment (reillyeon) [inline faraday/aprs.py]: `return sendAPRSPacket(socket, positionString)`

Here and below.

File: faraday/aprs.py [hunk 4]
```diff
@@ -693,14 +714,17 @@ def connectAPRSIS():
...
             except IOError as e:
+                # Close socket and setup for next connection attempt
                 logger.error(e)
+                aprssock.close()
+                aprssock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
 
             else:
                 logger.info("Connection successful!")
                 return aprssock
                 break
```
Comment (reillyeon) [inline faraday/aprs.py]: Not this change, but `break` after `return` does nothing.

File: faraday/aprs.py [hunk 5]
```diff
@@ -762,7 +786,8 @@ def main():
     """
 
     logger.info('Starting Faraday APRS-IS application')
-    sock = connectAPRSIS()
+    #sock = connectAPRSIS()
+    sock = ''
```
Comment (reillyeon) [inline faraday/aprs.py]: Why pass `sock` to `aprs_worker` at all if it calls `connectAPRSIS()` itself?

**Review by reillyeon — CHANGES_REQUESTED** [review]

File: faraday/aprs.py (hunk 1 above)
Comment (kb1lqc) [reply faraday/aprs.py]: Removed, `conn=False` now used in the initialization just above to force it to initialize on first loop. Thanks for catching this!

Comment (kb1lqc) [reply faraday/aprs.py]: ?

File: faraday/aprs.py (hunk 2 above)
Comment (kb1lqc) [reply faraday/aprs.py]: Removed

File: faraday/aprs.py (hunk 3 above)
Comment (kb1lqc) [reply faraday/aprs.py]: All `return status` lines replaced with obviously better `return sendAPRSPacket()`. Thanks for observing this unnecessary addition of variables!

File: faraday/aprs.py (hunk 4 above)
Comment (kb1lqc) [reply faraday/aprs.py]: Removed

File: faraday/aprs.py (hunk 5 above)
Comment (kb1lqc) [reply faraday/aprs.py]: Good point. I used to need this but with the current implementation of the socket connections I do not. Removed!

File: faraday/aprs.py [hunk 6]
```diff
@@ -693,14 +699,16 @@ def connectAPRSIS():
...
+                aprssock.close()
+                aprssock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
 
             else:
                 logger.info("Connection successful!")
                 return aprssock
-                break
 
-            sleep(10)  # Try to reconnect every 10 seconds
+            sleep(2)  # Try to reconnect every 10 seconds
```
Comment (reillyeon) [inline faraday/aprs.py]: Update the comment?

File: faraday/aprs.py [hunk 7]
```diff
@@ -268,11 +302,37 @@ def sendPositions(stations, socket):
...
+        b91c = base91.from_decimal(station["ADC3"], 2)
+        b91d = base91.from_decimal(station["ADC6"], 2)
+        b91e = base91.from_decimal(station["BOARDTEMP"], 2)
+        b91f = base91.from_decimal(ioList, 2)
+
+        b91Tlm = "|{0}{1}{2}{3}{4}{5}{6}|".format(b91seq, b91a, b91b, b91c, b91d, b91e, b91f)
+
+        # add telemetry to comments
+        comment = comment + b91Tlm
+        altComment = altComment + b91Tlm
```
Comment (reillyeon) [inline faraday/aprs.py]: comment += b91Tlm
    altComment += b91Tlm

`a += b` is more efficient than `a = a + b`.

**Review by reillyeon — APPROVED** [review]

File: faraday/aprs.py (hunk 7 above)
Comment (kb1lqc) [reply faraday/aprs.py]: Updated

File: faraday/aprs.py (hunk 6 above)
Comment (kb1lqc) [reply faraday/aprs.py]: Updated

**Comment by kb1lqc** [comment]
@reillyeon if you're going to help pull in the code that is effectively PR #269 then I should bring over the changelong here too. I appreciate the help on these if you are doing both!

**Comment by reillyeon** [comment]
PR #269 plus the changes here look good to me.

**Comment by kb1lqc** [comment]
Learn something new everyday... By pulling in #271 github automatically merged #269 since it was based on the same branch. This is awesome and in retrospect obvious! This means that as we progress through a big task we can commit mini PR's on it and still have one big PR at the end. If one PR that is a combination of a few smaller PR's in merged then we no longer have to worry about those. I like that!

@kb1lqd for awareness.

**Comment by kb1lqd** [comment]
Whoa thanks all and now I learned something new too, very cool PR trait!

**Pull request merged**